command/alloc_exec.go

```diff
@@ -0,0 +1,338 @@
+package command
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	"github.com/docker/docker/pkg/term"
+	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/api/contexts"
+	"github.com/posener/complete"
+)
+
+type AllocExecCommand struct {
+	Meta
+
+	Stdin  io.Reader
+	Stdout io.WriteCloser
+	Stderr io.WriteCloser
+}
+
+func (l *AllocExecCommand) Help() string {
+	helpText := `
+Usage: nomad alloc exec [options] <allocation> <command>
+
+  Run command inside the environment of the given allocation and task.
+
+General Options:
+
+  ` + generalOptionsUsage() + `
+
+Exec Specific Options:
+
+  -task <task-name>
+    Sets the task to exec command in
+
+  -job <job-id>
+    Use a random allocation from the specified job ID.
+
+  -i, --stdin=true
+    Pass stdin to the container, defaults to true
+
+  -t, --tty=true
+    Allocate a pseudo-tty, defaults to true if stdin is detected to be a tty session
+  `
+	return strings.TrimSpace(helpText)
+}
+
+func (l *AllocExecCommand) Synopsis() string {
+	return "Execute commands in task"
+}
+
+func (c *AllocExecCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"--task":  complete.PredictAnything,
+			"-job":    complete.PredictAnything,
+			"-t":      complete.PredictNothing,
+			"--tty":   complete.PredictNothing,
+			"--stdin": complete.PredictNothing,
+		})
+}
+
+func (l *AllocExecCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictFunc(func(a complete.Args) []string {
+		client, err := l.Meta.Client()
+		if err != nil {
+			return nil
+		}
+
+		resp, _, err := client.Search().PrefixSearch(a.Last, contexts.Allocs, nil)
+		if err != nil {
+			return []string{}
+		}
+		return resp.Matches[contexts.Allocs]
+	})
+}
+
+func (l *AllocExecCommand) Name() string { return "alloc exec" }
+
+func (l *AllocExecCommand) Run(args []string) int {
+	var job bool
+	var stdinShortOpt, stdinLongOpt, ttyShortOpt, ttyLongOpt bool
+	var task string
+
+	flags := l.Meta.FlagSet(l.Name(), FlagSetClient)
+	flags.Usage = func() { l.Ui.Output(l.Help()) }
+	flags.BoolVar(&job, "job", false, "")
```

Comment:
`job` should be a `StringVar` to allow it to be parsed out here, and to not enforce ordering of flags

Reply:
Like above, this matches `nomad logs` pattern and is used to discern whether command argument is an alloc or a job id.  If we make `job` take a string argument, we'll need to interpret args differently (to avoid requiring an alloc id) and complicates code a tad more and makes it not conform with other arguments.  If we revisit this, I'd prefer doing it in a follow up PR that changes other commands.